pyccl/halos/halo_model.py

```diff
@@ -137,6 +140,67 @@ def profile_norm(self, cosmo, a, prof):
         norm = 1. / self._integrate_over_mf(uk0)
         return norm
 
+    def number_counts(self, cosmo, sel, na=128, amin=None, amax=1.0):
+        """ Solves the integral:
+
+        .. math::
+            nc(sel) = \\int dM\\int da\\,\\frac{dV}{dad\\Omega}\\,n(M,a)\\,sel(M,a)
+
+        where :math:`n(M,a)` is the halo mass function, and
+        :math:`sel(M,a)` is the selection function as a function of halo mass
+        and scale factor.
+
+        Note that the selection function is normalized to integrate to unity and
+        assumed to represent the selection probaility per unit scale factor and
+        per unit mass.
+
+        Args:
+            cosmo (:class:`~pyccl.core.Cosmology`): a Cosmology object.
+            sel (callable): function of mass and scale factor that returns the
+                selection function. This function should take in floats or arrays
+                with a signature ``sel(m, a)`` and return an array with shape
+                ``(len(m), len(a))`` according to the numpy broadcasting rules.
+            na (int): number of samples in scale factor to be used in
+                the integrals. Default: 128.
+            amin (float): the minimum scale factor at which to start integrals
+                over the selection function.
+                Default: value of ``cosmo.cosmo.spline_params.A_SPLINE_MIN``
+            amax (float): the maximum scale factor at which to end integrals
+                over the selection function.
+                Default: 1.0
+
+        Returns:
+            float: the total number of clusters
+        """  # noqa
+
+        # get a values for integral
+        if amin is None:
+            amin = cosmo.cosmo.spline_params.A_SPLINE_MIN
+        a = np.linspace(amin, amax, na)
+
+        # compute the volume element
+        abs_dzda = 1 / a / a
+        dc = background.comoving_angular_distance(cosmo, a)
+        ez = background.h_over_h0(cosmo, a)
+        dh = physical_constants.CLIGHT_HMPC / cosmo['h']
+        dvdz = dh * dc**2 / ez
+        dvda = dvdz * abs_dzda
+
+        # now do m intergrals in a loop
+        mint = np.zeros_like(a)
+        for i, _a in enumerate(a):
+            self._get_ingredients(_a, cosmo, False)
+            _selm = np.atleast_2d(sel(self._mass, _a)).T
```

Comment:
Since you're transposing things here anyway, would it make sense to just ask that `sel` return arrays of shape ``(len(a), len(m))``?

Reply:
I was trying to match the dimensions of other quantities in the code under the assumption that this was useful. Happy to change.

Reply:
Ah wait, so are we doing this stupidly in other places too?

Reply:
(and by we I mean me...)

Reply:
yes

Reply:
Oh, actually, I remember there was a reason for this, to do with what's best to integrate profiles (which is a limiting factor often). OK, nevermind. Let's approve this.